### PR TITLE
Encode mongo connection strings as needed

### PR DESCRIPTION
**Pull request by wwlorey**
I didn't want to always encode connection strings just in case they're already encoded. So this only encodes if an encoding-related error is thrown first.

Fixes https://github.com/microsoft/vscode-cosmosdb/issues/1209

### CODE REVIEW

File: src/mongo/mongoConnectionStrings.ts [hunk 1]
```diff
@@ -49,6 +49,23 @@ export async function parseMongoConnectionString(connectionString: string): Prom
     let host: string;
     let port: string;
 
+    try {
+        // Attempt to connect to determine if the username/password need to be encoded
+        await connectToMongoClient(connectionString, appendExtensionUserAgent());
```
Comment (EricJizbaMSFT) [inline src/mongo/mongoConnectionStrings.ts:54]: Can you combine this with line 69? No need to connect twice for the happy-path.

```
let mongoClient: MongoClient;
try {
    mongoClient = await connectToMongoClient(connectionString, appendExtensionUserAgent());
} catch (error) {
    // unescaped stuff
    mongoClient = await connectToMongoClient(connectionString, appendExtensionUserAgent());
}
```

File: src/mongo/mongoConnectionStrings.ts [hunk 2]
```diff
@@ -49,6 +49,23 @@ export async function parseMongoConnectionString(connectionString: string): Prom
...
+    } catch (error) {
+        const parsedError: IParsedError = parseError(error);
+        if (parsedError.message.match(/unescaped/i)) {
+            const matches: RegExpMatchArray | null = connectionString.match(/^(.*):\/\/(.*):(.*)@(.*)/);
```
Comment (EricJizbaMSFT) [inline src/mongo/mongoConnectionStrings.ts:58]: How about pulling some of this stuff into it's own function and adding some unit tests? Other functions in this file already have it.

File: src/mongo/mongoConnectionStrings.ts [hunk 3]
```diff
@@ -49,6 +49,23 @@ export async function parseMongoConnectionString(connectionString: string): Prom
...
+            if (matches) {
+                const prefix: string = matches[1];
+                const username: string = matches[2];
+                const password: string = matches[3];
+                const hostAndQuery: string = matches[4];
+                connectionString = `${prefix}://${encodeURIComponent(username)}:${encodeURIComponent(password)}@${hostAndQuery}`;
```
Comment (EricJizbaMSFT) [inline src/mongo/mongoConnectionStrings.ts:64]: Nathan's recommendation in the issue was to escape with a slash, but you're using `encodeURIComponent` which is different. Do they both work?

Comment (wwlorey) [reply src/mongo/mongoConnectionStrings.ts:64]: I was not able to get escaping with a slash to work

**Pull request merged**